File: src/main/java/com/bigdata/spark/java/core/Persist.java
package com.bigdata.spark.java.core;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

/**
 * Created with IntelliJ IDEA.
 *
 * @Author: CuiCan
 * @Date: 2017-7-31
 * @Time: 15:08
 * @Description: RDD持久化
 */
public class Persist {

    public static void main(String[] args) {
        JavaSparkContext sc = new JavaSparkContext(
                new SparkConf().setAppName("persist").setMaster("local"));

        /**
         * cache()或者persist()的使用，是有规则的
         * 必须在transformation或者textFile等创建一个RDD之后，直接连续调用才可以。
         * 如果先创建RDD，再另起一行调用方法，会报错，大量的文件会丢失。
         */
        JavaRDD<String> lines = sc.textFile("E:\\Workspace\\intellij2017\\sparkgo\\src\\main\\resources\\txt\\spark.txt")
                .cache();

        long beginTime = System.currentTimeMillis();
        long count = lines.count();
        long endTime = System.currentTimeMillis();
        System.out.println("count = " + count);
        System.out.println("第一次耗时：" + (endTime - beginTime)+" ms.");

        beginTime = System.currentTimeMillis();
        count = lines.count();
        endTime = System.currentTimeMillis();
        System.out.println("count = " + count);
        System.out.println("第二次耗时：" + (endTime - beginTime)+" ms.");

        sc.close();
    }

}

File: src/main/scala/com/bigdata/spark/scala/core/Persist.scala
package com.bigdata.spark.scala.core

import org.apache.spark.storage.StorageLevel
import org.apache.spark.{SparkConf, SparkContext}

/**
  * Created with IntelliJ IDEA.
  *
  * Author: CuiCan
  * Date: 2017-7-31
  * Time: 16:14
  * Description: RDD持久化
  *              默认持久化策略：persist(StorageLevel.MEMORY_ONLY) = cache()
  */
object Persist {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("LineCount").setMaster("local")
    val sc = new SparkContext(conf)

    val rdd = sc.textFile("E:\\Workspace\\intellij2017\\sparkgo\\src\\main\\resources\\txt\\spark.txt")
                .persist(StorageLevel.MEMORY_ONLY)

    val startTime = System.currentTimeMillis()
    val count1 = rdd.count()
    val endTime = System.currentTimeMillis()
    println(count1)
    println("耗时一：" + (endTime - startTime) + " ms.")

    val beginTime = System.currentTimeMillis()
    val count2 = rdd.count()
    val overTime = System.currentTimeMillis()
    println(count2)
    println("耗时二：" + (overTime - beginTime) + " ms.")

  }
}
